File: leetcode2/2medium/최원준/Q300.java
package Leetcode.최원준;

/*
1. 아이디어 :
Bottom-Up DP를 사용하여, 각 숫자에 대해 가장 긴 증가하는 부분 수열의 길이를 계산합니다.

2. 시간복잡도 :
O( n^2 )

3. 자료구조/알고리즘 :
배열 / DP
 */

import java.util.Arrays;

public class Q300 {
    class Solution {
        public int lengthOfLIS(int[] nums) {
            int n = nums.length;
            int[] dp = new int[n];
            Arrays.fill(dp, 1);
            int ans = 0;
            for (int left = n-1; left>-1; left--) {
                for (int right = n-1; right > left; right--) {
                    if (nums[left] < nums[right]) {
                        dp[left] = Math.max(dp[left], dp[right] + 1);
                    }
                }
                ans = Math.max(ans, dp[left]);
                System.out.println(Arrays.toString(dp));
            }

            return ans;
        }
    }
}

File: leetcode2/2medium/노형민/Q300.cs
public class Solution
{
    public int LengthOfLIS(int[] nums)
    {
        int result = 0;
        int[] LISLength = new int[nums.Length];

        for (int i = 0; i < nums.Length; i++)
        {
            for (int j = 0; j < i; j++)
            {
                if (nums[i] > nums[j])
                {
                    LISLength[i] = Math.Max(LISLength[i], LISLength[j] + 1);
                }
            }

            if (LISLength[i] > result)
            {
                result = LISLength[i];
            }
        }

        //자기 자신이 포함되어야 하기에 + 1
        return result + 1;
    }
}